cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup.go

```diff
@@ -95,7 +96,21 @@ func (ng *nodegroup) DeleteNodes(nodes []*apiv1.Node) error {
 		if actualNodeGroup.Id() != ng.Id() {
 			return fmt.Errorf("node %q doesn't belong to node group %q", node.Spec.ProviderID, ng.Id())
 		}
+	}
+
+	// Step 2: if deleting len(nodes) would make the replica count
+	// <= 0, then the request to delete that many nodes is bogus
+	// and we fail fast.
+	replicas := ng.scalableResource.Replicas()
+
+	if replicas-int32(len(nodes)) <= 0 {
+		return fmt.Errorf("unable to delete %d machines in %q, machine replicas are <= 0 ", len(nodes), ng.Id())
+	}
 
+	// Step 3: annotate the corresponding machine that it is a
+	// suitable candidate for deletion and drop the replica count
+	// by 1. Fail fast on any error.
+	for _, node := range nodes {
```

Comment:
This will be definitely slower. Though, given we don't scale down by more than 1 in the test suite and we still don't guarantee any max. time for a nodegroup to be scaled down, it's acceptable.